lib/spack/spack/environment.py

```diff
@@ -1018,18 +1163,48 @@ def write(self):
         # invalidate _repo cache
         self._repo = None
 
-        # put the new user specs in the YAML
-        yaml_dict = config_dict(self.yaml)
-        yaml_spec_list = yaml_dict.setdefault('specs', [])
-        yaml_spec_list[:] = [str(s) for s in self.user_specs]
-
-        if self._view_path == self.default_view_path:
-            view = True
-        elif self._view_path:
-            view = self._view_path
+        # put any chagnes in the definitions in the YAML
+        for name, speclist in list(self.read_specs.items())[:-1]:
+            conf = config_dict(self.yaml)
+            active_yaml_lists = [l for l in conf.get('definitions', [])
+                                 if name in l and
+                                 _eval_conditional(l.get('when', 'True'))]
+
+            # Remove any specs in yaml that are not in internal representation
+            for ayl in active_yaml_lists:
+                # If it's not a string, it's a matrix. Those can't have changed
+                ayl[name][:] = [s for s in ayl.setdefault(name, [])
+                                if not isinstance(s, str) or
+                                Spec(s) in speclist.specs]
+
+            # Put the new specs into the first active list from the yaml
+            new_specs = [entry for entry in speclist.yaml_list
+                         if isinstance(entry, str) and
+                         not any(entry in ayl[name]
+                                 for ayl in active_yaml_lists)]
+            list_for_new_specs = active_yaml_lists[0].setdefault(name, [])
+            list_for_new_specs[:] = list_for_new_specs + new_specs
+
+        # put the new user specs in the YAML.
+        # This can be done directly because there can't be multiple definitions
+        # nor when clauses for `specs` list.
+        yaml_spec_list = config_dict(self.yaml).setdefault('specs', [])
+        yaml_spec_list[:] = self.user_specs.yaml_list
+
+        if self.views and len(self.views) == 1 and self.default_view_path:
```

Comment:
Can this stuff be consolidated somehow with the code in `_read_manifest`?  I'm a bit worried about the complexity here around dealing with `self.views` -- I think there needs to be a class or some methods wrapping this logic in one place so we don't lose track of it.